Update Cargo.toml

lveyde opened the pull request:
Usage of zlib backend for flate2 causes a compilation failure. One should use a rust_backend instead.

Otherwise a libz-sys crate compilation error similar to this will occur:

    Updating crates.io index
   Compiling libz-sys v1.1.12
    Checking priority-queue v1.3.2
The following warnings were emitted during compilation:

warning: error: unknown target triple 'wasm32-wasi', please use -triple or -arch

error: failed to run custom build command for `libz-sys v1.1.12`

Caused by:
  process didn't exit successfully: `/home/lveyde/git/rust/Create_a_Microservice_in_Rust/lveyde-a-stateful-microservice-lp/order_management/target/debug/build/libz-sys-0ed89978f8121fe7/build-script-build` (exit status: 1)
  --- stdout
  cargo:rerun-if-env-changed=LIBZ_SYS_STATIC
  cargo:rerun-if-changed=build.rs
  cargo:rerun-if-env-changed=ZLIB_NO_PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG_ALLOW_CROSS_wasm32-wasi
  cargo:rerun-if-env-changed=PKG_CONFIG_ALLOW_CROSS_wasm32_wasi
  cargo:rerun-if-env-changed=TARGET_PKG_CONFIG_ALLOW_CROSS
  cargo:rerun-if-env-changed=PKG_CONFIG_ALLOW_CROSS
  cargo:rerun-if-env-changed=PKG_CONFIG_wasm32-wasi
  cargo:rerun-if-env-changed=PKG_CONFIG_wasm32_wasi
  cargo:rerun-if-env-changed=TARGET_PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_wasm32-wasi
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_wasm32_wasi
  cargo:rerun-if-env-changed=TARGET_PKG_CONFIG_SYSROOT_DIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR
  cargo-warning=pkg-config has not been configured to support cross-compilation.

  Install a sysroot for the target platform and configure it via
  PKG_CONFIG_SYSROOT_DIR and PKG_CONFIG_PATH, or install a
  cross-compiling wrapper for pkg-config and set it via
  PKG_CONFIG environment variable.
  TARGET = Some("wasm32-wasi")
  OPT_LEVEL = Some("0")
  HOST = Some("x86_64-unknown-linux-gnu")
  cargo:rerun-if-env-changed=CC_wasm32-wasi
  CC_wasm32-wasi = None
  cargo:rerun-if-env-changed=CC_wasm32_wasi
  CC_wasm32_wasi = None
  cargo:rerun-if-env-changed=TARGET_CC
  TARGET_CC = None
  cargo:rerun-if-env-changed=CC
  CC = None
  cargo:rerun-if-env-changed=CRATE_CC_NO_DEFAULTS
  CRATE_CC_NO_DEFAULTS = None
  DEBUG = Some("true")
  cargo:rerun-if-env-changed=CFLAGS_wasm32-wasi
  CFLAGS_wasm32-wasi = None
  cargo:rerun-if-env-changed=CFLAGS_wasm32_wasi
  CFLAGS_wasm32_wasi = None
  cargo:rerun-if-env-changed=TARGET_CFLAGS
  TARGET_CFLAGS = None
  cargo:rerun-if-env-changed=CFLAGS
  CFLAGS = None
  running: "clang" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-fno-omit-frame-pointer" "--target=wasm32-wasi" "-I" "src/zlib" "-fvisibility=hidden" "-DZ_SOLO" "-DSTDC" "-D_LARGEFILE64_SOURCE" "-D_POSIX_SOURCE" "-o" "/home/lveyde/git/rust/Create_a_Microservice_in_Rust/lveyde-a-stateful-microservice-lp/order_management/target/wasm32-wasi/debug/build/libz-sys-7cf7184cfca6fdb6/out/lib/src/zlib/adler32.o" "-c" "src/zlib/adler32.c"
  cargo:warning=error: unknown target triple 'wasm32-wasi', please use -triple or -arch

  exit status: 1

  --- stderr


  error occurred: Command "clang" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-fno-omit-frame-pointer" "--target=wasm32-wasi" "-I" "src/zlib" "-fvisibility=hidden" "-DZ_SOLO" "-DSTDC" "-D_LARGEFILE64_SOURCE" "-D_POSIX_SOURCE" "-o" "/home/lveyde/git/rust/Create_a_Microservice_in_Rust/lveyde-a-stateful-microservice-lp/order_management/target/wasm32-wasi/debug/build/libz-sys-7cf7184cfca6fdb6/out/lib/src/zlib/adler32.o" "-c" "src/zlib/adler32.c" with args "clang" did not execute successfully (status code exit status: 1).


warning: build failed, waiting for other jobs to finish...

lveyde committed:
Update Cargo.toml

Usage of zlib backend for flate2 causes a compilation failure.
One should use a rust_backend instead.

Signed-off-by: Lev Veyde <lveyde@gmail.com>

lveyde commented:
Forced-pushed the same patch with Signed-off-by added, to make the DCO happy.

lveyde commented:
The issue seems to be related to this one: [issue 172](https://github.com/WebAssembly/wasi-sdk/issues/172)

Basically, on many platforms the llvm/clang doesn't support wasm32.

juntao commented:
Hmm @CaptainVincent can you help review this? Thanks.

lveyde committed:
Fixed link to u/s project

Signed-off-by: Lev Veyde <lveyde@gmail.com>

CaptainVincent commented:
@lveyde 
Hi, could you provide your build command and reproduce steps? 
Are you running `cargo build --target wasm32-wasi` under this repo? Or are you using this crate as a dependency in another repo and encountering errors? If so, could you also provide the repository where the build fails? 

Because I am unable to reproduce the issue locally and 
it looks like the CI for the zlib backend added last time is running well ([link](https://github.com/WasmEdge/mysql_async_wasi/actions/runs/5884167764/job/15958358136)).

I'll also give your PR a try later, but I'd like to confirm what issue we're addressing first. Thank you.

lveyde commented:
> @lveyde Hi, could you provide your build command and reproduce steps? Are you running `cargo build --target wasm32-wasi` under this repo? Or are you using this crate as a dependency in another repo and encountering errors? If so, could you also provide the repository where the build fails?
> 
> Because I am unable to reproduce the issue locally and it looks like the CI for the zlib backend added last time is running well ([link](https://github.com/WasmEdge/mysql_async_wasi/actions/runs/5884167764/job/15958358136)).
> 
> I'll also give your PR a try later, but I'd like to confirm what issue we're addressing first. Thank you.

Yes, certainly. Im using it as a dependency in my own project, and it gets compiled together with other dependency tree.

I'm using CentOS 7.9, and the main issue is that it's llvm/clang version doesn't support compilation into wasm targets.
Thus one can't compile any C code into Wasm on this distro (without using unsupported version of llvm/clang by i.e. compiling it from source).
That doesn't affect any native Rust code, as Rust comes with it's own compiler, but compiling any wrappers around C code will fail.

Requiring flate2/zlib as a direct dependency of mysql_async_wasi is causing the inclusion of libz-sys crate (which is a wrapper around zlib), even if one uses a feature option to choose a different configuration, i.e. "default", which normally was not supposed to include zlib.
And since the libz-sys crate doesn't contain a native Rust code, but rather C code, that as mentioned, causes a clang to fail to compile it into the required wasm32 target.

I think that using a rust_backend instead provides a good solution in this case.

CaptainVincent commented on Cargo.toml:
```diff
@@ -15,7 +15,7 @@ categories = ["asynchronous", "database"]
 bytes = "1.0"
 crossbeam = "0.8.1"
-flate2 = { version = "1.0", features = ["zlib"], default-features = false }
+flate2 = { version = "1.0", features = ["rust_backend"], default-features = false }
```
Thank your feedback. I agree that for switching to a platform-independent backend, but my previous approach was not ideal.

How about remove features here.
```
flate2 = { version = "1.0", default-features = false }
```

And we reorganize features section as below
```toml
[features]
common = [
    "mysql_common/bigdecimal03",
    "mysql_common/rust_decimal",
    "mysql_common/time03",
    "mysql_common/uuid",
    "mysql_common/frunk",
]
zlib = ["flate2/zlib"]
rust_backend = ["flate2/rust_backend"]
default = [
    "common",
    "rust_backend"
]
native-tls-tls = ["native-tls", "tokio-native-tls"]
rustls-tls = [
    "rustls",
    "tokio-rustls",
    "webpki",
    "webpki-roots",
    "rustls-pemfile",
]
default-rustls = [
    "common",
    "rust_backend",
    "rustls-tls",
]
nightly = []
minimal = ["zlib"]
```

And then, if the user has specific requirements, they can manually disable the default feature in `mysql_async_wasi` and select `common` and `zlib` to customize it as they desire again.

If you also agree and are willing to help with these adjustments, there are some topics related to features in the README that need to be synced. If it goes beyond your expectations and you prefer me to handle it, just let me know. I will take care of it as soon as possible.

--

BTW, I noticed that the TLS-related features are currently not compilable for a long time. 😿

lveyde replied:
That could work as well, will need to test it on my system.

And yes, I noticed that the README docs needs some cleanup and fixes regardless.

I haven't even gotten to begin to test the TLS...

lveyde replied:
Implemented the suggestions, as well as updated the README.md file.

lveyde replied:
@CaptainVincent can you please review?

lveyde replied:
Fixed both issues.

CaptainVincent reviewed (changes requested):
I add one suggestion.

lveyde committed:
Update Cargo.toml

Reworked the features, and updated the README.md file to reflect
the changes in the default configuration.

Also fixed some references to the original crate.

Signed-off-by: Lev Veyde <lveyde@gmail.com>

CaptainVincent commented on README.md:
```diff
@@ -18,47 +18,46 @@ mysql_async_wasi = "<desired version>"
-    is `flate2` backend). Enables:
+    is `zlib` `flate2` backend). Enables:
 
     -   `flate2/zlib"
```
-   `flate2/zlib`

Help this, thx.

CaptainVincent commented on README.md:
```diff
@@ -18,47 +18,46 @@ mysql_async_wasi = "<desired version>"
 
-    -   `native-tls-tls`
-    -   `flate2/zlib"
+    -   `flate2/rust_backend"
```
-   `flate2/rust_backend`

this too.

CaptainVincent reviewed (changes requested):
Two minor changes are needed. I've reviewed the rest part, and it looks fine.

CaptainVincent reviewed (approved):
LGTM. @juntao Could your help ship this.

juntao committed:
Update the crate version number for publication


Signed-off-by: Michael Yuan <michael@michaelyuan.com>

juntao commented:
I increased the minor version number so that I can publish this crate. Thank you!

lveyde commented:
> I increased the minor version number so that I can publish this crate. Thank you!

Sure, no problem!
